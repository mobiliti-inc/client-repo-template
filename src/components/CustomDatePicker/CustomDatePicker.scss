@import "~variables";

.container {
  font-family: $body-font;
  display: flex;
  flex-direction: column;
}

.container input {
  height: 50px;
  width: 100%;
  font-size: 1.25rem;
  color: $black;
  padding: 0;
  border: none;
  transition: all .5s;
  border-bottom: 1px solid rgba(29, 49, 85, 0.2);
  text-transform: capitalize;
  font-family: $body-font;

  &:focus {
    outline: none;
    border-bottom: 3px solid $blue;
  }

  &::-webkit-input-placeholder {
    color: $darker-blue;
    opacity: 0.6;
  }
}

.error-border input {
  border-bottom: 3px solid $fast-car-red;
}

.error {
  color: $fast-car-red;
  height: 14px;
  font-family: $hanken-sans;
  margin-top: 5px;
  font-size: 12px;
  line-height: 14px;
}

.icon {
  position: absolute;
  background-color: $white;
  top: 20px;
  width: 35px;
  border: 1px solid #d0d6de;
  border-radius: 5px;
  padding: 5px;
}

.icon:hover {
  background-color: rgba(204, 209, 218, .4);
}

.icon:active {
  background-color: $white;
}

.prev {
  left: 5px;
}

.next {
  right: 5px;
  -webkit-transform: rotate(180deg);
  /* Safari and Chrome */
  -moz-transform: rotate(180deg);
  /* Firefox */
  -ms-transform: rotate(180deg);
  /* IE 9 */
  -o-transform: rotate(180deg);
  /* Opera */
  transform: rotate(180deg);
}

.select {
  border: none;
  background: #2a98b1;
  color: #fff;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0 20px 0 5px;
  height: 40px;
  line-height: 30px;
  margin: 0 5px 0 0;
  border-radius: 2px;
  background-image: url(../../assets/icons/chevron-down-triangle.svg);
  background-position: right center;
  background-repeat: no-repeat;
}

.select:focus {
  outline: 0;
}

.calendar {
  display: flex;
  justify-content: center;
  margin-top: -9px;
}

.input-label {
  display: block;
  font-size: 12px;
  transform: translateY(-60px);
  color: $darker-blue;
  opacity: 0.6;
  font-family: $hanken-sans;
  line-height: 14px;
  text-transform: capitalize;
  transition: all .5s;

  &--focused {
    opacity: 0;
    visibility: hidden;
    transform: translateY(-40px);
  }
}
