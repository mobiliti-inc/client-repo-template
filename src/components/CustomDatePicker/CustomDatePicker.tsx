/* eslint no-underscore-dangle: 0 */
import React from 'react';
import moment from 'moment';
import classNames from 'classnames';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import styles from './CustomDatePicker.scss';
import { directionArrow } from '../../assets/icons';

type CustomDatePickerProps = {
	placeholder?: string,
	isValid?: boolean,
	errorMessage?: string,
	onDatePicked: (...args: any[]) => any,
	openToDate?: string,
	license?: boolean,
	selectedDate?: string,
	label?: string,
	id?: string,
	numberOfYears?: number,
	isDayBlocked?: (...args: any[]) => any,
	isOutsideRange?: (...args: any[]) => any,
	negative?: boolean
};

type CustomDatePickerState = {
	date: null,
	valid: boolean,
	error: string,
	focused: boolean
};

class CustomDatePicker extends React.PureComponent<CustomDatePickerProps, CustomDatePickerState> {
	state = {
		date: null,
		valid: true,
		error: '',
		focused: false
	};

	componentDidMount() {
		const { selectedDate } = this.props;
		if (selectedDate && selectedDate._isValid) this.setDate(selectedDate);
	}

	componentDidUpdate = (prevProps, prevState) => {
		const { openToDate, selectedDate } = this.props;
		const { focused, date } = this.state;
		if (focused && prevState.focused !== focused && openToDate && this.formatDate(openToDate)._isValid && !date) {
			this.setDate(openToDate);
		}
		if (selectedDate && prevProps.selectedDate !== selectedDate && this.formatDate(selectedDate)._isValid) {
			this.setDate(selectedDate);
		}
	}

	onDateChange = date => {
		const { onDatePicked } = this.props;
		if (date && date._isValid) {
			onDatePicked({ value: moment(date).format('MM/DD/YYYY'), error: '', valid: true });
			this.setState({ date, valid: true, error: '' });
		} else {
			onDatePicked();
			this.setState({ valid: false, error: 'Please enter a valid date' });
		}
	}

	setDate = date => this.setState({ date: this.formatDate(date), valid: true, error: '' });

	// Combine classes
	getClasses = () => classNames('container', { 'error-border': !this.state.valid || !this.props.isValid });

	formatDate = date => moment(date, 'MM/DD/YYYY');

	renderMonthElement = ({ month, onMonthSelect, onYearSelect }) => (
		<div className={styles.calendar}>
			<div>
				<select value={month.month()} onChange={e => onMonthSelect(month, e.target.value)} className={styles.select}>
					{moment.months().map((label, value) => (
						<option value={value} key={label}>
							{label}
						</option>
					))}
				</select>
			</div>
			<div>
				<select value={month.year()} onChange={e => onYearSelect(month, e.target.value)} className={styles.select}>
					{[...Array(this.props.numberOfYears)].map((value, index) => {
						const indexValue = moment().year() + index - (this.props.negative ? this.props.numberOfYears : 0);
						return (
							<option value={indexValue} key={indexValue}>
								{indexValue}
							</option>
						);
					})}
					{this.props.license &&
						[...Array(10)].map((value, index) => {
							const indexValue = moment().year() + index;
							return (
								<option value={indexValue} key={indexValue}>
									{indexValue}
								</option>
							);
						})}
				</select>
			</div>
		</div>
	)

	renderErrorBox = () => {
		const { error, valid } = this.state;
		const { errorMessage, isValid } = this.props;
		if (!valid || !isValid) {
			if (errorMessage && errorMessage.length > 0) {
				return <span styleName="error"> {errorMessage} </span>;
			}
			return <span styleName="error"> {error} </span>;
		}
		return null;
	}

	render() {
		const { placeholder, label, id, errorMessage } = this.props;
		const { date, error } = this.state;
		const navPrevIcon = <img src={directionArrow} alt="" styleName="icon prev" />;
		const navNextIcon = <img src={directionArrow} alt="" styleName="icon next" />;
		return (
			<div styleName={this.getClasses()}>
				<SingleDatePicker
					id={id}
					date={date}
					numberOfMonths={1}
					daySize={35}
					onDateChange={this.onDateChange}
					focused={this.state.focused}
					navPrev={navPrevIcon}
					navNext={navNextIcon}
					noBorder
					block
					isOutsideRange={this.props.isOutsideRange}
					hideKeyboardShortcutsPanel
					placeholder={placeholder}
					horizontalMonthPadding={5}
					onFocusChange={({ focused }) => this.setState({ focused })}
					renderMonthElement={this.renderMonthElement}
					isDayBlocked={this.props.isDayBlocked}
				/>
				{this.renderErrorBox()}
				{label && (
					<label styleName={classNames('input-label', { 'input-label--focused': this.state.focused || !date || error || errorMessage })} htmlFor={id}>
						{label}
					</label>
				)}
			</div>
		);
	}
}

export default CSSModules(CustomDatePicker, styles, { allowMultiple: true });
