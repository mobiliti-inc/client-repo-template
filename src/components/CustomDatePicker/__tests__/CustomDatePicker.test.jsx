import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import CustomDatePicker from '../CustomDatePicker';

describe('CustomDatePicker', () => {
	const tree = shallow(<CustomDatePicker onDatePicked={() => {}} />);

	it('renders', () => {
		matchSnapshot(tree);
	});

	it('has a datepicker component', () => {
		expect(tree.find(SingleDatePicker)).toHaveLength(1);
	});

	it('should be able to pick a different date', () => {
		const dateToSet = moment('1999-09-09');
		tree.instance().onDateChange(dateToSet);
		expect(tree.state().date).toEqual(dateToSet);
	});
});
