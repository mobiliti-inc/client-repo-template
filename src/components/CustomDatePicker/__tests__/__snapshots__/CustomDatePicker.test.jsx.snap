// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDatePicker renders 1`] = `
<div
  className="container"
>
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    appendToBody={false}
    block={true}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={null}
    daySize={35}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={true}
    horizontalMargin={0}
    horizontalMonthPadding={5}
    id=""
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={
      <img
        alt=""
        className="icon next"
        src="test-file-stub"
      />
    }
    navPrev={
      <img
        alt=""
        className="icon prev"
        src="test-file-stub"
      />
    }
    noBorder={true}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder={null}
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={[Function]}
    renderMonthText={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
